File: producer-api-python-rest-lambda/service/event_processing.py
"""Event processing service."""

import json
import logging
from datetime import datetime
from typing import Optional

from constants import API_NAME
from models.event import Event, EntityUpdate
from repository import BusinessEventRepository, EntityRepository, get_connection_pool

logger = logging.getLogger(__name__)

# Entity type to table name mapping
ENTITY_TABLE_MAP = {
    "Car": "car_entities",
    "Loan": "loan_entities",
    "LoanPayment": "loan_payment_entities",
    "ServiceRecord": "service_record_entities",
}


class EventProcessingService:
    """Service for processing events."""
    
    def __init__(self, business_event_repo: BusinessEventRepository, pool):
        """Initialize service with repositories."""
        self.business_event_repo = business_event_repo
        self.pool = pool
        self._persisted_event_count = 0
    
    def _log_persisted_event_count(self):
        """Log persisted event count every 10 events."""
        self._persisted_event_count += 1
        if self._persisted_event_count % 10 == 0:
            logger.info(f"{API_NAME} *** Persisted events count: {self._persisted_event_count} ***")
    
    def _get_entity_repository(self, entity_type: str) -> Optional[EntityRepository]:
        """Get the appropriate entity repository for the entity type."""
        table_name = ENTITY_TABLE_MAP.get(entity_type)
        if table_name is None:
            logger.warning(f"{API_NAME} Unknown entity type: {entity_type}")
            return None
        return EntityRepository(self.pool, table_name)
    
    async def process_event(self, event: Event) -> None:
        """Process a single event."""
        logger.info(f"{API_NAME} Processing event: {event.event_header.event_name}")
        
        # 1. Save entire event to business_events table
        await self.save_business_event(event)
        
        # 2. Extract and save entities to their respective tables
        for entity_update in event.event_body.entities:
            await self.process_entity_update(entity_update)
        
        self._log_persisted_event_count()
    
    async def save_business_event(self, event: Event) -> None:
        """Save the entire event to business_events table."""
        event_id = event.event_header.uuid or f"event-{datetime.utcnow().isoformat()}"
        event_name = event.event_header.event_name
        event_type = event.event_header.event_type
        created_date = event.event_header.created_date or datetime.utcnow()
        saved_date = event.event_header.saved_date or datetime.utcnow()
        
        # Convert event to dict for JSONB storage
        event_data = event.model_dump(mode='json')
        
        await self.business_event_repo.create(
            event_id=event_id,
            event_name=event_name,
            event_type=event_type,
            created_date=created_date,
            saved_date=saved_date,
            event_data=event_data,
        )
        logger.info(f"{API_NAME} Successfully saved business event: {event_id}")
    
    async def process_entity_update(self, entity_update: EntityUpdate) -> None:
        """Process a single entity update."""
        logger.info(
            f"{API_NAME} Processing entity creation for type: {entity_update.entity_type} "
            f"and id: {entity_update.entity_id}"
        )
        
        entity_repo = self._get_entity_repository(entity_update.entity_type)
        if entity_repo is None:
            logger.warning(
                f"{API_NAME} Skipping entity with unknown type: {entity_update.entity_type}"
            )
            return
        
        exists = await entity_repo.exists_by_entity_id(entity_update.entity_id)
        
        if exists:
            logger.warning(
                f"{API_NAME} Entity already exists, updating: {entity_update.entity_id}"
            )
            await self.update_existing_entity(entity_repo, entity_update)
        else:
            logger.info(
                f"{API_NAME} Entity does not exist, creating new: {entity_update.entity_id}"
            )
            await self.create_new_entity(entity_repo, entity_update)
    
    async def create_new_entity(
        self, entity_repo: EntityRepository, entity_update: EntityUpdate
    ) -> None:
        """Create a new entity."""
        entity_id = entity_update.entity_id
        entity_type = entity_update.entity_type
        now = datetime.utcnow()
        
        # Extract entity data from updated_attributes
        entity_data = entity_update.updated_attributes.copy() if isinstance(
            entity_update.updated_attributes, dict
        ) else {}
        
        # Remove entityHeader from entity_data if it exists (nested structure)
        entity_header = entity_data.pop("entityHeader", None) or entity_data.pop("entity_header", None)
        
        # Extract createdAt and updatedAt from entityHeader if present, otherwise from entity_data, otherwise use now
        if entity_header and isinstance(entity_header, dict):
            created_at_str = entity_header.get("createdAt") or entity_header.get("created_at")
            updated_at_str = entity_header.get("updatedAt") or entity_header.get("updated_at")
        else:
            created_at_str = entity_data.pop("createdAt", None) or entity_data.pop("created_at", None)
            updated_at_str = entity_data.pop("updatedAt", None) or entity_data.pop("updated_at", None)
        
        # Parse datetime strings to datetime objects
        created_at = self._parse_datetime(created_at_str) if created_at_str else now
        updated_at = self._parse_datetime(updated_at_str) if updated_at_str else now
        
        await entity_repo.create(
            entity_id=entity_id,
            entity_type=entity_type,
            created_at=created_at,
            updated_at=updated_at,
            entity_data=entity_data,
        )
        logger.info(f"{API_NAME} Successfully created entity: {entity_id}")
    
    def _parse_datetime(self, dt_str: str) -> datetime:
        """Parse datetime string to datetime object."""
        if isinstance(dt_str, datetime):
            return dt_str
        
        try:
            # Try ISO format
            return datetime.fromisoformat(dt_str.replace('Z', '+00:00'))
        except (ValueError, AttributeError):
            try:
                # Try parsing with dateutil
                from dateutil import parser
                return parser.parse(dt_str)
            except Exception:
                return datetime.utcnow()
    
    async def update_existing_entity(
        self, entity_repo: EntityRepository, entity_update: EntityUpdate
    ) -> None:
        """Update an existing entity."""
        entity_id = entity_update.entity_id
        updated_at = datetime.utcnow()
        
        # Extract entity data from updated_attributes
        entity_data = entity_update.updated_attributes.copy() if isinstance(
            entity_update.updated_attributes, dict
        ) else {}
        
        # Remove entityHeader from entity_data if it exists
        entity_data.pop("entityHeader", None)
        entity_data.pop("entity_header", None)
        
        # Remove entityHeader fields that might be at top level
        entity_data.pop("createdAt", None)
        entity_data.pop("created_at", None)
        entity_data.pop("updatedAt", None)
        entity_data.pop("updated_at", None)
        
        await entity_repo.update(
            entity_id=entity_id,
            updated_at=updated_at,
            entity_data=entity_data,
        )
        logger.info(f"{API_NAME} Successfully updated entity: {entity_id}")

File: producer-api-go-rest-lambda/internal/service/event_processing.go
package service

import (
	"context"
	"encoding/json"
	"fmt"
	"producer-api-go-rest-lambda/internal/models"
	"producer-api-go-rest-lambda/internal/repository"
	"strconv"
	"sync/atomic"
	"time"

	"producer-api-go-rest-lambda/internal/constants"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"
	"go.uber.org/zap"
)

// Entity type to table name mapping
var entityTableMap = map[string]string{
	"Car":           "car_entities",
	"Loan":          "loan_entities",
	"LoanPayment":   "loan_payment_entities",
	"ServiceRecord": "service_record_entities",
}

type EventProcessingService struct {
	businessEventRepo   *repository.BusinessEventRepository
	eventHeaderRepo     *repository.EventHeaderRepository
	pool                *pgxpool.Pool
	persistedEventCount *atomic.Uint64
	logger              *zap.Logger
}

func NewEventProcessingService(
	businessEventRepo *repository.BusinessEventRepository,
	pool *pgxpool.Pool,
	logger *zap.Logger,
) *EventProcessingService {
	return &EventProcessingService{
		businessEventRepo:   businessEventRepo,
		eventHeaderRepo:     repository.NewEventHeaderRepository(pool),
		pool:                pool,
		persistedEventCount: &atomic.Uint64{},
		logger:              logger,
	}
}

func (s *EventProcessingService) logPersistedEventCount() {
	count := s.persistedEventCount.Add(1)
	if count%10 == 0 {
		s.logger.Info(fmt.Sprintf("%s *** Persisted events count", constants.APIName()), zap.Uint64("count", count))
	}
}

func (s *EventProcessingService) getEntityRepository(entityType string) *repository.EntityRepository {
	tableName, ok := entityTableMap[entityType]
	if !ok {
		s.logger.Warn(fmt.Sprintf("%s Unknown entity type: %s", constants.APIName(), entityType))
		return nil
	}
	return repository.NewEntityRepository(s.pool, tableName)
}

// ProcessEvent processes a single event within a transaction.
// All database operations (business_events, event_headers, entities) are executed
// atomically within a single transaction. If any operation fails, all changes are rolled back.
func (s *EventProcessingService) ProcessEvent(ctx context.Context, event *models.Event) error {
	s.logger.Info(fmt.Sprintf("%s Processing event: %s", constants.APIName(), event.EventHeader.EventName))

	// Generate event ID if not provided
	eventID := event.EventHeader.UUID
	if eventID == nil || *eventID == "" {
		id := fmt.Sprintf("event-%s", time.Now().UTC().Format(time.RFC3339Nano))
		eventID = &id
	}

	// Begin transaction
	tx, err := s.pool.Begin(ctx)
	if err != nil {
		return fmt.Errorf("failed to begin transaction: %w", err)
	}
	defer tx.Rollback(ctx) // Rollback if not committed

	// 1. Save entire event to business_events table
	if err := s.saveBusinessEvent(ctx, event, *eventID, tx); err != nil {
		return fmt.Errorf("failed to save business event: %w", err)
	}

	// 2. Save event header to event_headers table
	if err := s.saveEventHeader(ctx, event, *eventID, tx); err != nil {
		return fmt.Errorf("failed to save event header: %w", err)
	}

	// 3. Extract and save entities to their respective tables
	for _, entityUpdate := range event.Entities {
		if err := s.processEntityUpdate(ctx, entityUpdate, *eventID, tx); err != nil {
			return fmt.Errorf("failed to process entity update: %w", err)
		}
	}

	// Commit transaction
	if err := tx.Commit(ctx); err != nil {
		return fmt.Errorf("failed to commit transaction: %w", err)
	}

	s.logPersistedEventCount()
	s.logger.Info(fmt.Sprintf("%s Successfully processed event in transaction: %s", constants.APIName(), *eventID))
	return nil
}

func (s *EventProcessingService) saveBusinessEvent(
	ctx context.Context,
	event *models.Event,
	eventID string,
	tx pgx.Tx,
) error {
	eventName := event.EventHeader.EventName
	eventType := event.EventHeader.EventType
	createdDate := event.EventHeader.CreatedDate
	savedDate := event.EventHeader.SavedDate

	// Use current time if dates are not provided
	now := time.Now()
	if createdDate == nil {
		createdDate = &now
	}
	if savedDate == nil {
		savedDate = &now
	}

	// Convert event to map for JSONB storage
	eventJSON, err := json.Marshal(event)
	if err != nil {
		return fmt.Errorf("failed to marshal event: %w", err)
	}

	var eventData map[string]interface{}
	if err := json.Unmarshal(eventJSON, &eventData); err != nil {
		return fmt.Errorf("failed to unmarshal event data: %w", err)
	}

	err = s.businessEventRepo.Create(ctx, eventID, eventName, eventType, createdDate, savedDate, eventData, tx)
	if err != nil {
		return err
	}

	s.logger.Info(fmt.Sprintf("%s Successfully saved business event: %s", constants.APIName(), eventID))
	return nil
}

func (s *EventProcessingService) saveEventHeader(
	ctx context.Context,
	event *models.Event,
	eventID string,
	tx pgx.Tx,
) error {
	eventName := event.EventHeader.EventName
	eventType := event.EventHeader.EventType
	createdDate := event.EventHeader.CreatedDate
	savedDate := event.EventHeader.SavedDate

	// Use current time if dates are not provided
	now := time.Now()
	if createdDate == nil {
		createdDate = &now
	}
	if savedDate == nil {
		savedDate = &now
	}

	// Convert eventHeader to map for JSONB storage
	headerJSON, err := json.Marshal(event.EventHeader)
	if err != nil {
		return fmt.Errorf("failed to marshal event header: %w", err)
	}

	var headerData map[string]interface{}
	if err := json.Unmarshal(headerJSON, &headerData); err != nil {
		return fmt.Errorf("failed to unmarshal header data: %w", err)
	}

	err = s.eventHeaderRepo.Create(ctx, eventID, eventName, eventType, createdDate, savedDate, headerData, tx)
	if err != nil {
		return err
	}

	s.logger.Info(fmt.Sprintf("%s Successfully saved event header: %s", constants.APIName(), eventID))
	return nil
}

func (s *EventProcessingService) processEntityUpdate(
	ctx context.Context,
	entity models.Entity,
	eventID string,
	tx pgx.Tx,
) error {
	s.logger.Info(fmt.Sprintf("%s Processing entity for type: %s and id: %s",
		constants.APIName(), entity.EntityHeader.EntityType, entity.EntityHeader.EntityID))

	entityRepo := s.getEntityRepository(entity.EntityHeader.EntityType)
	if entityRepo == nil {
		s.logger.Warn(fmt.Sprintf("%s Skipping entity with unknown type: %s",
			constants.APIName(), entity.EntityHeader.EntityType))
		return nil
	}

	exists, err := entityRepo.ExistsByEntityID(ctx, entity.EntityHeader.EntityID, tx)
	if err != nil {
		return fmt.Errorf("failed to check if entity exists: %w", err)
	}

	if exists {
		s.logger.Warn(fmt.Sprintf("%s Entity already exists, updating: %s",
			constants.APIName(), entity.EntityHeader.EntityID))
		return s.updateExistingEntity(ctx, entityRepo, entityUpdate, eventID, tx)
	} else {
		s.logger.Info(fmt.Sprintf("%s Entity does not exist, creating new: %s",
			constants.APIName(), entity.EntityHeader.EntityID))
		return s.createNewEntity(ctx, entityRepo, entityUpdate, eventID, tx)
	}
}

func (s *EventProcessingService) createNewEntity(
	ctx context.Context,
	entityRepo *repository.EntityRepository,
	entity models.Entity,
	eventID string,
	tx pgx.Tx,
) error {
	entityID := entity.EntityHeader.EntityID
	entityType := entity.EntityHeader.EntityType
	now := time.Now()

	// Extract entity data from entity properties
	entityData := entity.Properties

	// Use createdAt and updatedAt from entityHeader
	createdAt := &entity.EntityHeader.CreatedAt
	updatedAt := &entity.EntityHeader.UpdatedAt

	err := entityRepo.Create(ctx, entityID, entityType, createdAt, updatedAt, entityData, &eventID, tx)
	if err != nil {
		return fmt.Errorf("failed to create entity: %w", err)
	}

	s.logger.Info(fmt.Sprintf("%s Successfully created entity: %s", constants.APIName(), entityID))
	return nil
}

func (s *EventProcessingService) updateExistingEntity(
	ctx context.Context,
	entityRepo *repository.EntityRepository,
	entity models.Entity,
	eventID string,
	tx pgx.Tx,
) error {
	entityID := entity.EntityHeader.EntityID
	updatedAt := time.Now()

	// Extract entity data from entity properties
	entityData := entity.Properties

	// Remove entityHeader fields that might be at top level
	delete(entityData, "createdAt")
	delete(entityData, "created_at")
	delete(entityData, "updatedAt")
	delete(entityData, "updated_at")

	err := entityRepo.Update(ctx, entityID, updatedAt, entityData, &eventID, tx)
	if err != nil {
		return fmt.Errorf("failed to update entity: %w", err)
	}

	s.logger.Info(fmt.Sprintf("%s Successfully updated entity: %s", constants.APIName(), entityID))
	return nil
}

func (s *EventProcessingService) parseDateTime(dtStr string) (time.Time, error) {
	// Try ISO 8601 formats
	formats := []string{
		time.RFC3339,
		time.RFC3339Nano,
		"2006-01-02T15:04:05Z",
		"2006-01-02T15:04:05.000Z",
		"2006-01-02T15:04:05Z07:00",
	}

	for _, format := range formats {
		if t, err := time.Parse(format, dtStr); err == nil {
			return t, nil
		}
	}

	// Try parsing as Unix timestamp (milliseconds)
	if ms, err := strconv.ParseInt(dtStr, 10, 64); err == nil {
		secs := ms / 1000
		nsecs := (ms % 1000) * 1000000
		return time.Unix(secs, nsecs), nil
	}

	return time.Now(), fmt.Errorf("unable to parse datetime string: %s", dtStr)
}
